File: app/src/main/java/io/plan8/backoffice/adapter/TaskViewAdapter.java
package io.plan8.backoffice.adapter;

import android.databinding.BindingAdapter;
import android.graphics.drawable.GradientDrawable;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.AlphaAnimation;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.RelativeLayout;

import com.prolificinteractive.materialcalendarview.CalendarDay;
import com.prolificinteractive.materialcalendarview.MaterialCalendarView;
import com.prolificinteractive.materialcalendarview.format.TitleFormatter;

import io.plan8.backoffice.Constants;
import io.plan8.backoffice.R;
import io.plan8.backoffice.view.BlurView;
import io.plan8.backoffice.view.Plan8TaskCalendarView;

/**
 * Created by SSozi on 2017. 11. 28..
 */

public class TaskViewAdapter {
    @BindingAdapter("taskViewAdapter:displayCalendar")
    public static void setDisplayCalendar(final Plan8TaskCalendarView view, Boolean isOpenedCalendar) {
        Animation slideDownAnimation;
        Animation slideUpAnimation;

        if (!view.isAlreadyInflated()) {
            slideDownAnimation = null;
            slideUpAnimation = null;
            view.setSelectedDate(CalendarDay.today());
            view.setAlreadyInflated(true);
            view.setVisibility(View.INVISIBLE);
            view.setPagingEnabled(false);
            return;
        }

        slideDownAnimation = AnimationUtils.loadAnimation(view.getContext(), R.anim.slide_down);
        slideDownAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                view.setVisibility(View.VISIBLE);
                view.setPagingEnabled(true);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });

        slideUpAnimation = AnimationUtils.loadAnimation(view.getContext(), R.anim.slide_up);
        slideUpAnimation.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                view.setVisibility(View.INVISIBLE);
                view.setPagingEnabled(false);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });

        if (isOpenedCalendar) {
            view.startAnimation(slideDownAnimation);
        } else {
            view.startAnimation(slideUpAnimation);
        }
    }

    @BindingAdapter("taskViewAdapter:initCalendar")
    public static void initCalendar(MaterialCalendarView view, Boolean isOpenedCalendar) {
        view.setTitleFormatter(new TitleFormatter() {
            @Override
            public CharSequence format(CalendarDay day) {
                return "" + day.getYear() + "년 " + (day.getMonth() + 1) + "월";
            }
        });
    }

    @BindingAdapter("taskViewAdapter:setReadStatus")
    public static void setReadStatus(RelativeLayout view, String taskStatus) {
        view.setBackgroundResource(R.drawable.circle);
        GradientDrawable bgShape = (GradientDrawable) view.getBackground();

        if (taskStatus.equals(Constants.TASK_STATUS_BLUE)) {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.taskStatusBlue));
        } else if (taskStatus.equals(Constants.TASK_STATUS_ORANGE)) {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.taskStatusOrange));
        } else if (taskStatus.equals(Constants.TASK_STATUS_RED)) {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.taskStatusRed));
        } else {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.transparent));
        }
    }

    @BindingAdapter("taskViewAdapter:setCloseStatus")
    public static void setCloseStatus(RelativeLayout view, String taskStatus) {
        view.setBackgroundResource(R.drawable.circle);
        GradientDrawable bgShape = (GradientDrawable) view.getBackground();
        if (taskStatus == Constants.TASK_STATUS_GREEN) {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.taskStatusGreen));
        } else {
            bgShape.setColor(ContextCompat.getColor(view.getContext(), R.color.transparent));
        }
    }

    @BindingAdapter("taskViewAdapter:fadeout")
    public static void fadeout(final BlurView view, Boolean display) {
        if (!view.isAlreadyInflated()) {
            view.setAlreadyInflated(true);
            view.setVisibility(View.GONE);
            return;
        }
        if (display) {
            view.setVisibility(View.VISIBLE);
        } else {
            AlphaAnimation animation = new AlphaAnimation(1.0f, 0.0f);
            animation.setDuration(200);
            animation.setInterpolator(new AccelerateDecelerateInterpolator());
            animation.setAnimationListener(new Animation.AnimationListener() {
                @Override
                public void onAnimationStart(Animation animation) {

                }

                @Override
                public void onAnimationEnd(Animation animation) {
                    view.setVisibility(View.GONE);
                }

                @Override
                public void onAnimationRepeat(Animation animation) {

                }
            });
            view.setVisibility(View.VISIBLE);
            view.startAnimation(animation);
        }
    }
}

File: app/src/main/java/io/plan8/backoffice/adapter/TaskViewAdapter.kt
package io.plan8.backoffice.adapter

import android.databinding.BindingAdapter
import android.graphics.drawable.GradientDrawable
import android.support.v4.content.ContextCompat
import android.util.Log
import android.view.DragEvent
import android.view.MotionEvent
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator
import android.view.animation.AlphaAnimation
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.RelativeLayout
import android.widget.Toast
import com.prolificinteractive.materialcalendarview.CalendarDay
import com.prolificinteractive.materialcalendarview.MaterialCalendarView
import io.plan8.backoffice.Constants
import io.plan8.backoffice.R
import io.plan8.backoffice.view.BlurView
import io.plan8.backoffice.view.Plan8TaskCalendarView
import kotlinx.android.synthetic.main.activity_main.view.*


/**
 * Created by chokwanghwan on 2017. 11. 3..
 */
class TaskViewAdapter {
    companion object {
        @BindingAdapter("taskViewAdapter:displayCalendar")
        @JvmStatic
        fun setDisplayCalendar(view: Plan8TaskCalendarView, isOpenedCalendar: Boolean) {
            if (!view.isAlreadyInflated) {
                view.selectedDate = CalendarDay.today()
                view.isAlreadyInflated = true
                view.visibility = View.INVISIBLE
                view.isPagingEnabled = false
                return
            }
            var slideDownAnimation: Animation? = null
            var slideUpAnimation: Animation? = null

            if (slideDownAnimation == null || slideUpAnimation == null) {
                slideDownAnimation = AnimationUtils.loadAnimation(view.context, R.anim.slide_down)
                slideDownAnimation!!.setAnimationListener(object : Animation.AnimationListener {
                    override fun onAnimationStart(animation: Animation) {}

                    override fun onAnimationRepeat(animation: Animation) {}

                    override fun onAnimationEnd(animation: Animation) {
                        view.visibility = View.VISIBLE
                        view.isPagingEnabled = true
                    }
                })

                slideUpAnimation = AnimationUtils.loadAnimation(view.context, R.anim.slide_up)
                slideUpAnimation!!.setAnimationListener(object : Animation.AnimationListener {
                    override fun onAnimationStart(animation: Animation) {}

                    override fun onAnimationRepeat(animation: Animation) {}

                    override fun onAnimationEnd(animation: Animation) {
                        view.visibility = View.INVISIBLE
                        view.isPagingEnabled = false
                    }
                })
            }

            if (isOpenedCalendar) {
                view.startAnimation(slideDownAnimation)
            } else {
                view.startAnimation(slideUpAnimation)
            }
        }

        @BindingAdapter("taskViewAdapter:initCalendar")
        @JvmStatic
        fun initCalendar(view: MaterialCalendarView, isOpenedCalendar: Boolean) {
            view.setTitleFormatter({ day: CalendarDay? -> "" + day!!.year + "년 " + (day.month + 1) + "월" })
        }

        @BindingAdapter("taskViewAdapter:setStatus")
        @JvmStatic
        fun setStatus(view: RelativeLayout, taskStatus: String) {
            view.setBackgroundResource(R.drawable.circle)
            val bgShape = view.background as GradientDrawable
            when (taskStatus) {
                Constants.TASK_STATUS_BLUE -> bgShape.setColor(ContextCompat.getColor(view.context, R.color.taskStatusBlue))
                Constants.TASK_STATUS_ORANGE -> bgShape.setColor(ContextCompat.getColor(view.context, R.color.taskStatusOrange))
                else -> bgShape.setColor(ContextCompat.getColor(view.context, R.color.taskStatusRed))
            }
        }

        @BindingAdapter("taskViewAdapter:fadeout")
        @JvmStatic
        fun fadeout(view: BlurView, display: Boolean) {
            if (!view.isAlreadyInflated) {
                view.isAlreadyInflated = true
                view.visibility = View.GONE
                return
            }
            if (display) {
                view.visibility = View.VISIBLE
            } else {
                val animation = AlphaAnimation(1.0f, 0.0f)
                animation.duration = 200
                animation.interpolator = AccelerateDecelerateInterpolator()
                animation.setAnimationListener(object : Animation.AnimationListener {
                    override fun onAnimationStart(animation: Animation) {

                    }

                    override fun onAnimationEnd(animation: Animation) {
                        view.visibility = View.GONE
                    }

                    override fun onAnimationRepeat(animation: Animation) {

                    }
                })
                view.visibility = View.VISIBLE
                view.startAnimation(animation)
            }
        }
    }
}